pkg/util/crud.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2018 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// https://github.com/kubernetes-sigs/controller-runtime/blob/6100e07/pkg/controller/controllerutil/controllerutil.go#L117
+//
+// This file contains a fork of the above function. At the time of writing the
+// latest release of controller-runtime is v0.1.10, which contains a buggy
+// CreateOrUpdate implementation.
+
+// CreateOrUpdate creates or updates the given object in the Kubernetes
+// cluster. The object's desired state must be reconciled with the existing
+// state inside the passed in callback MutateFn. The MutateFn is called
+// regardless of creating or updating an object.
+func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, f MutateFn) error {
```

Comment:
Nice, perhaps we should create an issue to track this removal after we upgrade to 0.2+

Reply:
I didn't capture in my comment the three changes I made to this function:

* I removed the `OperationResult` return value simply because we did not use it.
* I removed a test of whether the namespace changed. My reason for removing this test is now defunct as I found a better way to deal with the issue I was trying to address.
* I added some context to the errors with `errors.Wrap()`.

So all that said, I do think we should return to the upstream implementation and should have a JIRA issue to track that. I'd be sad to lose the additional error context though - perhaps I could submit a small PR to controller-runtime to see whether they would be interested in having that context upstream.

Reply:
https://github.com/crossplaneio/crossplane/issues/426